Lib/threading.py

```diff
@@ -1094,11 +1094,24 @@ def _wait_for_tstate_lock(self, block=True, timeout=-1):
         # If the lock is acquired, the C code is done, and self._stop() is
         # called.  That sets ._is_stopped to True, and ._tstate_lock to None.
         lock = self._tstate_lock
-        if lock is None:  # already determined that the C code is done
+        if lock is None:
+            # already determined that the C code is done
             assert self._is_stopped
-        elif lock.acquire(block, timeout):
-            lock.release()
-            self._stop()
+            return
+
+        try:
+            if lock.acquire(block, timeout):
+                lock.release()
+                self._stop()
+        except:
+            if lock.locked():
+                # bpo-45274: lock.acquire() acquired the lock, but the function
+                # was interrupted with an exception before reaching the
+                # lock.release(). It can happen if a signal handler raises an
+                # exception, like CTRL+C which raises KeyboardInterrupt.
+                lock.release()
```

Comment:
I don't think that this code is 100% reliable if lock.release() gets interrupted by a **second** exception (ex: raised by a second signal: fatality!). Maybe a context manager could be used. But I'm exhausted, I will think about this code after a good night :-)

Reply:
I _believe_ we've still got bugs in the context manager exit implementation where even it can miss something in this scenario and not call `__exit__`.  https://bugs.python.org/issue29988
This came up during the dev sprint in 2017 when working on the interpreter loop.

So this PR, while not technically a fix, is at least an improvement and should help the single Ctrl-C KeyboardInterrupt case.

Reply:
The whole function should be reimplemented in C to better control how signals are handled.

In pure Python, I don't think that it's possible to fully control handle any possible exception at any line number.

@serhiy-storchaka proposed to rewrite acquire()+release() in C to make sure that at least the lock remains consistent: https://bugs.python.org/issue45274#msg402532 So it doesn't handle exceptions in the _stop() method.

Reply:
But I only wanted to enhance the code, I'm not interested to rewrite threading.Thread in C.